pkg/netpol/ingressanalyzer/ingress_analyzer.go

```diff
@@ -0,0 +1,195 @@
+	ia.servicesToPeersMap[svc.Namespace][svc.Name] = peers
+	return nil
+}
+
+func (ia *IngressAnalyzer) getServicePeers(svc *corev1.Service) ([]eval.Peer, error) {
+	svcStr := types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace}.String()
+	if svc.Spec.Selector == nil {
+		return nil, errors.New(scan.Service + " " + svcStr + " " + missingSelectorError)
+	}
+	svcLabelsSelector, err := convertServiceSelectorToLabelSelector(svc.Spec.Selector, svcStr)
+	if err != nil {
+		return nil, err
+	}
+	return ia.pe.GetSelectedPeers(svcLabelsSelector, svc.Namespace), nil
+}
+
+// utility func
+func convertServiceSelectorToLabelSelector(svcSelector map[string]string, svcStr string) (labels.Selector, error) {
+	labelsSelector := metav1.LabelSelector{MatchLabels: svcSelector}
+	selectorRes, err := metav1.LabelSelectorAsSelector(&labelsSelector)
+	if err != nil {
+		return nil, errors.New(scan.Service + " " + svcStr + " " + selectorError)
+	}
+	return selectorRes, nil
+}
+
+// //////////////////////////////////////////////////////////////////////////////////////////////
+// routes analysis
+
+const (
+	allowedTargetKind  = scan.Service
+	routeTargetKindErr = "target kind error"
+	routeBackendsErr   = "alternate backends error"
+)
+
+// this function populates routesToServicesMap
+func (ia *IngressAnalyzer) mapRouteToServices(rt *ocroutev1.Route) error {
+	services, err := getRouteServices(rt)
+	if err != nil {
+		return err
+	}
+	if _, ok := ia.routesToServicesMap[rt.Namespace]; !ok {
+		ia.routesToServicesMap[rt.Namespace] = make(map[string][]string)
+	}
+	ia.routesToServicesMap[rt.Namespace][rt.Name] = services
+	return nil
+}
+
+func getRouteServices(rt *ocroutev1.Route) ([]string, error) {
+	routeStr := types.NamespacedName{Namespace: rt.Namespace, Name: rt.Name}.String()
+	// Currently, only 'Service' is allowed as the kind of target that the route is referring to.
+	if rt.Spec.To.Kind != "" && rt.Spec.To.Kind != allowedTargetKind {
+		return nil, errors.New(scan.Route + " " + routeStr + ": " + routeTargetKindErr)
+	}
+
+	targetServices := make([]string, len(rt.Spec.AlternateBackends)+1)
+	targetServices[0] = rt.Spec.To.Name
+	for i, backend := range rt.Spec.AlternateBackends {
+		if backend.Kind != "" && backend.Kind != allowedTargetKind {
+			return nil, errors.New(scan.Route + " " + routeStr + ": " + routeBackendsErr)
+		}
+		targetServices[i+1] = backend.Name
+	}
+	return targetServices, nil
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+// Ingress allowed conns
+
+// AllowedIngressConnections returns map of the allowed external ingress-controller's connections of each targeted peer
+func (ia *IngressAnalyzer) AllowedIngressConnections() map[string]eval.Connection {
+	// if there is at least one route/ ingress object that targets a service which selects a dst peer,
+	// then we have an ingress conns to the peer
+
+	// get all targeted peer pods
+	targetedPeersSet := make(map[eval.Peer]bool, 0)
```

Comment:
is it safe to use a map where the keys are pointers to workload objects? the function `GetSelectedPeers` returns new objects on each call...

Reply:
true, I'll make it a map from workload_name to workload

Reply:
maybe you can only have one map `map[string]eval.Connection` built for the result, while collecting all target peers.

Reply:
done